### PR TITLE
Search projects by name

### PR DESCRIPTION
**Pull request by AnnieJovitha**

### CODE REVIEW

File: ui/lib/screens/projects/projects_screen.dart [hunk 1]
```diff
@@ -20,24 +21,31 @@ class ProjectsScreen extends StatefulWidget {
 class _ProjectsScreenState extends State<ProjectsScreen> {
   late ProjectService service;
   late Future<List<Project>> projects;
+  String _filter = '';
```
Comment (timovandeput) [inline ui/lib/screens/projects/projects_screen.dart:24]: Flutter (Dart) convention is to use `var` where possible. (The empty value allows the compiler to infer the type here.)

File: ui/lib/screens/projects/projects_screen.dart [hunk 2]
```diff
@@ -3,6 +3,7 @@
  * SPDX-License-Identifier: MIT
  */
 
+import 'package:bom_bar_ui/screens/packages/name_filter.dart';
```
Comment (timovandeput) [inline ui/lib/screens/projects/projects_screen.dart:6]: I try to use relative imports for anything inside the project, and absolute imports for packages from outside the project.

PS: Test code needs to always use absolute imports, since it lives outside the `lib` directory.

File: ui/lib/screens/projects/projects_screen.dart [hunk 3]
```diff
@@ -20,24 +21,31 @@ class ProjectsScreen extends StatefulWidget {
...
+            _filter = fragment;
+            projects = service.findProjects(_filter);
+          }),
+        ),
         actions: [
           IconButton(
             icon: Icon(Icons.refresh),
             onPressed: () => setState(() {
-              projects = service.allProjects();
+              projects = service.findProjects(_filter);
```
Comment (timovandeput) [inline ui/lib/screens/projects/projects_screen.dart:48]: I don't think reload still makes a lot of sense in combination with the search bar. (I added it because there previously was no way to update the list.)

Let's remove the action button here (the packages screen also has no reload button).

PS: I think you then also don't need to track the `_filter` value anymore.

File: ui/lib/services/bom_bar_client.dart [hunk 4]
```diff
@@ -38,9 +38,10 @@ class BomBarClient {
     ));
   }
 
-  /// Returns all available projects.
-  Future<List<Project>> getProjects() async {
-    final response = await dio.getUri(_projectsUrl);
+  /// Returns all or projects matching the provided [filter] fragment.
+  Future<List<Project>> findProjectsBySearchFragment(String? filter) async {
```
Comment (timovandeput) [inline ui/lib/services/bom_bar_client.dart:42]: Dart has a syntax to pass an optional parameter, which seems more fitting here: `foo([String? filter])`.

(You can then call this function without parameter to indicate there is no fragment.)

Comment (timovandeput) [inline ui/lib/services/bom_bar_client.dart:42]: I am missing a test for the `null` value of `filter`.

(My expectation is that this would request projects without passing a `q` parameter or perhaps an empty string.)

Comment (timovandeput) [inline ui/lib/services/bom_bar_client.dart:42]: This function is "searching by search fragment", but then accepting a parameter called `filter`.

Suggestion is to name the function e.g. `findProjects()` and naming the parameter something like "fragment".

File: ui/lib/services/bom_bar_client.dart [hunk 5]
```diff
@@ -38,9 +38,10 @@ class BomBarClient {
     ));
   }
 
-  /// Returns all available projects.
-  Future<List<Project>> getProjects() async {
-    final response = await dio.getUri(_projectsUrl);
+  /// Returns all or projects matching the provided [filter] fragment.
```
Comment (timovandeput) [inline ui/lib/services/bom_bar_client.dart:41]: Textual detail: "all or projects"? :-)

File: ui/test/services/bom_bar_client_test.dart [hunk 6]
```diff
@@ -24,27 +24,45 @@ void main() {
...
+        test('queries for all projects with no search fragment', () async {
           server.respondJson({
             'results': [
-              {'id': projectId}
+              {'id': projectId, 'title': 'ProjectA'},
+              {'id': projectId, 'title': 'ProjectB'}
             ]
           });
-          final projects = await client.getProjects();
+          final projects = await client.findProjectsBySearchFragment('');
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:35]: You are passing a fragment here: the empty string.

Suggestion is to pass no parameter (=`null` value) instead.

File: ui/test/services/bom_bar_client_test.dart [hunk 7]
```diff
@@ -24,27 +24,45 @@ void main() {
...
+          final request = server.requests.first;
+          expect(request.method, 'GET');
+          expect(request.path,
+              BomBarClient.baseUrl.resolve('/projects/').toString());
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:39]: Are you sure there is no `q` query parameter in this case?

File: ui/test/services/bom_bar_client_test.dart [hunk 8]
```diff
@@ -24,27 +24,45 @@ void main() {
...
       const projectId = 'projectId';
 
-      group('Get all projects', () {
-        test('queries projects', () async {
+      group('Get projects based on search query', () {
+        test('queries for all projects with no search fragment', () async {
           server.respondJson({
             'results': [
-              {'id': projectId}
+              {'id': projectId, 'title': 'ProjectA'},
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:31]: Why return more than one project? (Does that add value to the test?)

PS: All projects now have the same id, and are therefore actually the same project. :-)

File: ui/test/services/bom_bar_client_test.dart [hunk 9]
```diff
@@ -24,27 +24,45 @@ void main() {
     group('Projects', () {
       const projectId = 'projectId';
 
-      group('Get all projects', () {
-        test('queries projects', () async {
+      group('Get projects based on search query', () {
+        test('queries for all projects with no search fragment', () async {
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:28]: I am trying to stick to the "Arrange - Act - Assert" structure of unit tests, leaving a blank line between the sections.

Please also add the blank lines in new tests, so they match the rest of the tests.

File: ui/test/services/bom_bar_client_test.dart [hunk 10]
```diff
@@ -24,27 +24,45 @@ void main() {
...
+        });
 
+        test('queries for projects matching search fragment', () async {
+          server.respondJson({
+            'results': [
+              {'id': projectId, 'title': 'ProjectA'}
+            ]
+          });
+          final projects =
+              await client.findProjectsBySearchFragment('ProjectA');
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:51]: Suggestion is to create a (local) `const` for the fragment, and use it in the Act and Assert statements of this test.

(It might also be clearer if you use a value like "fragment" for that constant.)

File: ui/test/services/bom_bar_client_test.dart [hunk 11]
```diff
@@ -24,27 +24,45 @@ void main() {
...
+          expect(request.queryParameters, {'q': 'ProjectA'});
           expect(projects.length, 1);
           expect(projects[0].id, projectId);
         });
 
         test('throws for server error status', () {
           server.respondStatus(404);
 
-          expect(client.getProjects(), throwsA(isInstanceOf<DioError>()));
+          expect(client.findProjectsBySearchFragment(''),
```
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:64]: If the parameter becomes optional, then you could remove the empty string here.

File: ui/test/services/project_service_test.dart [hunk 12]
```diff
@@ -36,20 +36,32 @@ void main() {
     });
 
     group('Projects query', () {
-      test('queries list of projects', () async {
-        when(client.getProjects())
-            .thenAnswer((_) => Future.value([Project(id: projectId)]));
+      test('query to list all projects when no filter applied', () async {
+        when(client.findProjectsBySearchFragment(''))
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart:40]: Suggestion is to pass a fragment (instead of the empty string), so you the test shows this is what is passed between this call and the REST client.

File: ui/test/services/project_service_test.dart [hunk 13]
```diff
@@ -36,20 +36,32 @@ void main() {
...
+            .thenAnswer((_) => Future.value([
+                  Project(id: projectId, title: 'ProjectA'),
+                  Project(id: projectId, title: 'ProjectB')
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart:43]: Limit the response to a single project to limit the cognitive distraction. (The list of projects is a detail for this test.)

File: ui/test/services/bom_bar_client_test.dart (hunk 8 above)
Comment (timovandeput) [inline ui/test/services/bom_bar_client_test.dart:31]: Why add a title to the projects in the new tests? (This is not a required parameter and it serves no purpose in the tests.)

File: ui/test/services/project_service_test.dart [hunk 14]
```diff
@@ -36,20 +36,32 @@ void main() {
...
+                ]));
 
-        final projects = await service.allProjects();
+        final projects = await service.findProjects('');
+
+        expect(projects, hasLength(2));
+      });
+
+      test('query to list only one filtered project - ProjectB', () async {
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart:51]: I try to name tests after the expected behaviour of the object under test, which leads to names like e.g. "queries by fragment".

I don't think "ProjectB" is relevant to the behaviour of the service, so my suggestion is to therefore not mention it in the title.

File: ui/test/services/project_service_test.dart [hunk 15]
```diff
@@ -36,20 +36,32 @@ void main() {
...
+        when(client.findProjectsBySearchFragment('ProjectB')).thenAnswer(
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart:52]: Suggestion to move the fragment parameter to a (local) `const` declaration, and assign it a descriptive value like "fragment".

File: ui/test/services/project_service_test.dart [hunk 16]
```diff
@@ -36,20 +36,32 @@ void main() {
...
+            (_) => Future.value([Project(id: projectId, title: 'ProjectB')]));
+
+        final projects = await service.findProjects('ProjectB');
 
         expect(projects, hasLength(1));
       });
 
       test('throws for list projects failure', () {
-        when(client.getProjects())
+        when(client.findProjectsBySearchFragment(''))
```
Comment (timovandeput) [inline ui/test/services/project_service_test.dart:61]: Suggestion to change to a parameter-less invocation.

**Review by timovandeput — CHANGES_REQUESTED** [review]

File: ui/lib/services/project_service.dart [hunk 17]
```diff
@@ -52,6 +45,14 @@ class ProjectService {
     });
   }
 
+  /// Returns all projects that match the [fragment] with their (PURL-based) reference.
+  /// Supports "projects_fragment/name_fragment" as fragment specification.
+  Future<List<Project>> findProjects([String? fragment]) => _execute(() async {
+        final projects = await _client.findProjectsBySearchFragment(fragment);
+        log('Searching projects matching "$fragment"');
```
Comment (timovandeput) [inline ui/lib/services/project_service.dart:52]: When not providing a fragment, this will log _Searching projects matching "null"_

Perhaps somehow indicate explicitly that no fragment was provided?

File: ui/lib/services/bom_bar_client.dart [hunk 18]
```diff
@@ -38,9 +38,10 @@ class BomBarClient {
     ));
   }
 
-  /// Returns all available projects.
-  Future<List<Project>> getProjects() async {
-    final response = await dio.getUri(_projectsUrl);
+  /// Returns all projects or projects matching the provided [fragment] fragment.
```
Comment (timovandeput) [inline ui/lib/services/bom_bar_client.dart:41]: The last "fragment" can be dropped, as the generated help nicely keeps the `[fragment]` text in the description.

File: ui/lib/services/bom_bar_client.dart [hunk 19]
```diff
@@ -38,9 +38,10 @@ class BomBarClient {
...
+  Future<List<Project>> findProjectsBySearchFragment([String? fragment]) async {
```
Comment (timovandeput) [inline ui/lib/services/bom_bar_client.dart:42]: Wondering if we should also rename this method to `findProjects(...)`, as the search fragment seems obvious from the parameter. (And if no fragment is provided, it still finds project.)

**Review by timovandeput — CHANGES_REQUESTED** [review]
(Some remaining minor details.)

**Review by timovandeput — APPROVED** [review]
